File: supabase/functions/play-ai-turn/_game/board.ts
// Kelimeki — tahta yardımcıları (saf fonksiyonlar)
//
// src/utils/board.ts'in KOPYASI — bkz. types.ts'teki not.
import { SIZE } from './constants.ts';
import type { CellKey, Tile } from './types.ts';

export type Board = (Tile | null)[][];
export type Placed = Record<CellKey, Tile>;

export function key(r: number, c: number): CellKey {
  return `${r},${c}`;
}

export function inBounds(r: number, c: number): boolean {
  return r >= 0 && r < SIZE && c >= 0 && c < SIZE;
}

/** Bir taşın görünen harfini döndürür (joker ise seçilen harf). */
export function tileLetter(tile: Tile): string {
  return tile.wild ? tile.wildLetter ?? '' : tile.letter;
}

function getLetterAt(board: Board, placed: Placed, r: number, c: number): string | null {
  const p = placed[key(r, c)];
  if (p) return tileLetter(p);
  const b = board[r]?.[c];
  if (b) return tileLetter(b);
  return null;
}

export interface FormedWord {
  word: string;
  coords: [number, number][];
}

function fullWordWithCoords(
  board: Board,
  placed: Placed,
  r: number,
  c: number,
  dr: number,
  dc: number,
): FormedWord {
  let sr = r;
  let sc = c;
  while (inBounds(sr - dr, sc - dc) && getLetterAt(board, placed, sr - dr, sc - dc)) {
    sr -= dr;
    sc -= dc;
  }
  let word = '';
  const coords: [number, number][] = [];
  let rr = sr;
  let rc = sc;
  while (inBounds(rr, rc) && getLetterAt(board, placed, rr, rc)) {
    word += getLetterAt(board, placed, rr, rc);
    coords.push([rr, rc]);
    rr += dr;
    rc += dc;
  }
  return { word, coords };
}

/**
 * Bu turdaki yerleştirmelerle oluşan tüm kelimeleri (ana + çapraz)
 * koordinatlarıyla birlikte döndürür. 2 harften kısa kelimeler atlanır.
 */
export function getFormedWords(board: Board, placed: Placed): FormedWord[] {
  const coords = Object.keys(placed).map((k) => k.split(',').map(Number) as [number, number]);
  if (coords.length === 0) return [];

  const rows = [...new Set(coords.map((p) => p[0]))];
  const horiz = rows.length === 1;

  const seen = new Set<string>();
  const result: FormedWord[] = [];

  const addWord = (fw: FormedWord) => {
    if (fw.word.length < 2) return;
    const [er, ec] = fw.coords[fw.coords.length - 1];
    const sig = `${fw.coords[0][0]},${fw.coords[0][1]}-${er},${ec}`;
    if (seen.has(sig)) return;
    seen.add(sig);
    result.push(fw);
  };

  addWord(
    fullWordWithCoords(board, placed, coords[0][0], coords[0][1], horiz ? 0 : 1, horiz ? 1 : 0),
  );

  for (const [r, c] of coords) {
    const [cdr, cdc] = horiz ? [1, 0] : [0, 1];
    addWord(fullWordWithCoords(board, placed, r, c, cdr, cdc));
  }

  return result;
}

File: mobile/kelimeki_core/lib/src/rules/board.dart
// Kelimeki core — tahta yardımcıları (src/utils/board.ts portu).
import '../constants.dart';
import '../model/game_state.dart' show Board, Placed;
import '../model/results.dart';
import '../model/tile.dart';
import '../model/types.dart';

Board createEmptyBoard() =>
    List.generate(boardSize, (_) => List<Tile?>.filled(boardSize, null, growable: false));

String cellKey(int r, int c) => '$r,$c';

/// "r,c" anahtarını (r, c) koordinatına çözer.
Cell parseKey(String k) {
  final i = k.indexOf(',');
  return (int.parse(k.substring(0, i)), int.parse(k.substring(i + 1)));
}

bool _inBounds(int r, int c) => r >= 0 && r < boardSize && c >= 0 && c < boardSize;

/// Bir taşın görünen harfi (joker ise seçilen harf; seçilmemişse boş dize).
String tileLetter(Tile tile) => tile.wild ? (tile.wildLetter ?? '') : tile.letter;

String? _getLetterAt(Board board, Placed placed, int r, int c) {
  final p = placed[cellKey(r, c)];
  if (p != null) return tileLetter(p);
  final b = _inBounds(r, c) ? board[r][c] : null;
  if (b != null) return tileLetter(b);
  return null;
}

/// (r,c)'den (dr,dc) yönünde uzanan tam kelime — koordinatlarıyla.
FormedWord _fullWordWithCoords(
    Board board, Placed placed, int r, int c, int dr, int dc) {
  var sr = r, sc = c;
  while (_inBounds(sr - dr, sc - dc) &&
      _getLetterAt(board, placed, sr - dr, sc - dc) != null) {
    sr -= dr;
    sc -= dc;
  }
  final buf = StringBuffer();
  final coords = <Cell>[];
  var rr = sr, rc = sc;
  while (_inBounds(rr, rc) && _getLetterAt(board, placed, rr, rc) != null) {
    buf.write(_getLetterAt(board, placed, rr, rc));
    coords.add((rr, rc));
    rr += dr;
    rc += dc;
  }
  return FormedWord(word: buf.toString(), coords: coords);
}

/// (r,c) hücresinden (dr,dc) yönünde uzanan tam kelime — web
/// `getFullWordAt` (src/utils/board.ts) eşleniği. Anlam sorgusu için tahtadaki
/// bir taşa dokunulduğunda o hücreden geçen yatay/dikey kelimeyi bulur.
/// Mevcut özel yardımcıyı sarar; davranış değişmez (golden vector paritesi).
String fullWordAt(Board board, Placed placed, int r, int c, int dr, int dc) =>
    _fullWordWithCoords(board, placed, r, c, dr, dc).word;

/// Bu turdaki yerleştirmelerle oluşan tüm kelimeler (ana + çapraz).
/// 2 harften kısa kelimeler atlanır. Sıra TS ile birebir: önce ana kelime,
/// sonra her yerleştirme için çapraz kelime (placed ekleme sırasında).
List<FormedWord> getFormedWords(Board board, Placed placed) {
  final coords = [for (final k in placed.keys) parseKey(k)];
  if (coords.isEmpty) return [];

  final rows = {for (final p in coords) p.$1};
  final horiz = rows.length == 1;

  final seen = <String>{};
  final result = <FormedWord>[];

  void addWord(FormedWord fw) {
    if (fw.word.length < 2) return;
    // Başlangıç + bitiş hücresi imzası: aynı hücreden başlayan yatay/dikey
    // iki kelime ayrı sayılmalı (TS'teki aynı imza).
    final e = fw.coords.last;
    final s = fw.coords.first;
    final sig = '${s.$1},${s.$2}-${e.$1},${e.$2}';
    if (seen.contains(sig)) return;
    seen.add(sig);
    result.add(fw);
  }

  // Ana kelime (yerleştirme yönünde).
  addWord(_fullWordWithCoords(board, placed, coords[0].$1, coords[0].$2,
      horiz ? 0 : 1, horiz ? 1 : 0));

  // Her yerleştirme için çapraz kelime.
  for (final p in coords) {
    final cdr = horiz ? 1 : 0;
    final cdc = horiz ? 0 : 1;
    addWord(_fullWordWithCoords(board, placed, p.$1, p.$2, cdr, cdc));
  }

  return result;
}
